### PR TITLE
IMR restore feature implementation

### PR DESCRIPTION
**Pull request by plbossart**
improvement of https://github.com/thesofproject/linux/pull/2989 with my own comments fixed my way. No functionality change compared to PR #2989

### CODE REVIEW

File: sound/soc/sof/intel/hda-loader.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-loader.c:29]: you missed the "slave" in the commit message :)

Comment (plbossart) [reply sound/soc/sof/intel/hda-loader.c:29]: darn, fixed now. Also added my Signed-off-by on stuff I modified.

**Review by ranj063 — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]

**Review by keyonjie — APPROVED** [review]
Thanks, lgtm.

**Comment by keyonjie** [comment]
@plbossart @marc-hb better to run a more stressful test before merging.

**Comment by keqiaozhang** [comment]
> @plbossart @marc-hb better to run a more stressful test before merging.

I had triggered the stress testing, test report ID is 8998.
The results are good and no regressions were found.
Test recipe: kernel [PR3340](https://github.com/thesofproject/linux/pull/3340) + SOF [PR 4893](https://github.com/thesofproject/sof/pull/4893)

**Comment by RanderWang** [comment]
currently IMR is enabled by fw_ready message which is not applicable to IPC4, so IPC4 can't be enabled by this way.

**Comment by ranj063** [comment]
> currently IMR is enabled by fw_ready message which is not applicable to IPC4, so IPC4 can't be enabled by this way.

@RanderWang for IPC4, shouldn't this be a default feature? For SOF we need the fw ready message because older versions do not support it.

**Comment by kv2019i** [comment]
Due to the complicated history of this feature, I'll let @plbossart you do the merge.

**Comment by RanderWang** [comment]
> > currently IMR is enabled by fw_ready message which is not applicable to IPC4, so IPC4 can't be enabled by this way.
> 
> @RanderWang for IPC4, shouldn't this be a default feature? For SOF we need the fw ready message because older versions do not support it.

yes, It should be a default feature for IPC4 but current code can't enabled it for IPC4 since it checks fw_ready info. We may need another method for IPC4

**Comment by plbossart** [comment]
I'd like to re-run a test with https://github.com/thesofproject/sof/pull/5109 before merging. I'll do this tomorrow.

**Comment by plbossart** [comment]
SOFCI TEST

**Pull request merged**
